### PR TITLE
Optimized allocations by using an IntArray instead of a mutableList

### PR DESCRIPTION
**Pull request by G4E**
Hey ! 
I forgot to send you this one. 
It greatly reduce memory pressure and garbage collection, as we don't create new lists on each draw.
You don't want a mutable list, as the list itself won't change after being created. However, each value in the list/array is still mutable. Like that, we only allocate the Int once and we mutate them when needed.
Bonus, the app use less memory overall !

### CODE REVIEW

File: app/src/main/java/rak/pixellwp/cycling/models/Palette.kt [hunk 1]
```diff
@@ -10,31 +10,31 @@ import kotlin.math.min
...
     fun cycle(timePassed: Int) {
         //it's important we copy the base values as each time we cycle it 'starts from 0'; it's not additive
-        colors = baseColors.toMutableList()
-        cycles
-                .filter { it.rate != 0 }
-                .forEach { cycle ->
+        baseColors.copyInto(colors)
+        cycles.forEach { cycle ->
+                    if(cycle.rate==0)
+                        return@forEach //Similar to break in a conventionnal loop
```
Comment (ManApart) [inline app/src/main/java/rak/pixellwp/cycling/models/Palette.kt:29]: I probably won't be able to run this locally until friday or this weekend.
Why not use filter? Is that not available on the intArray or is this somehow more performant?

Comment (G4E) [reply app/src/main/java/rak/pixellwp/cycling/models/Palette.kt:29]: No worries, we are not in a rush !
I have to admit, it's my C training speaking there : like that we have to iterate only once on the array, instead of needing to iterate first to keep only the non zero values, and then iterate again to apply the color's algorithm.
Moreover, i'm pretty sure it prevents the creation of an intermediate array, containing only the non zero values.

Comment (ManApart) [reply app/src/main/java/rak/pixellwp/cycling/models/Palette.kt:29]: To my knowledge, the `.filter` _would_ create an intermediate list of the cycle and _would_ iterate a second time over those cycles that are not a 0 rate. However, I don't think that's much overhead given the cycle list is usually very small. (I may remember wrong but I'd guess 10 or less on average), so I'm not sure it's a large performance change.

That said, I think the return@forEach is like you said, the same as a 'break' in java, but I think what you'd need here is a 'continue'. I believe as written, this change would prevent any further cycles being evaluated after finding one cycle with a 0 rate. Since the cycles are not sorted, this could mean valid cycles are prevented.

If both my understandings are correct, I'd say it's worth keeping the filter for clarity since I don't think it will hurt performance without having so many cycles that another performance optimization would be better anyway.

**Review by ManApart — APPROVED** [review]

**Comment by ManApart** [comment]
I've stalled out on moving this to fdroid. If you have any interesting in pushing it along, that would be great, but no pressure. Just wanted to let you know in case you wanted to help move it over. Thanks again for the optimization work!

https://gitlab.com/fdroid/rfp/-/issues/2004#note_825866181

File: app/src/main/java/rak/pixellwp/cycling/models/Palette.kt (hunk 1 above)
Comment (G4E) [reply app/src/main/java/rak/pixellwp/cycling/models/Palette.kt:29]: Yeah i'm sorry, i wanted to say 'continue' and not 'break', I goofed on that one. I can confirm that it indeed works like 'continue'.
I'll profile with and without that notation to confirm if it is useful or just an overly enthusiastic optimization !

**Comment by G4E** [comment]
I already looked into that, and I will invest a bit more time into it, but like you I didn't understand what the issue actually is :/

File: app/src/main/java/rak/pixellwp/cycling/models/Palette.kt (hunk 1 above)
Comment (G4E) [reply app/src/main/java/rak/pixellwp/cycling/models/Palette.kt:29]: https://user-images.githubusercontent.com/10716685/175832101-a2c5c760-e5f0-4689-bd38-3bcde257dfd3.mp4
So I profiled both algorithms (with filter, and without filter+return@forEach) :
Sadly, my pc was dying while using the emulator + the profiler + the screencap, so there are not much fps here ...
With filter :
https://user-images.githubusercontent.com/10716685/175832145-49fbad02-9953-433f-8c26-2eb7f26a1e91.mp4

Without filter :
https://user-images.githubusercontent.com/10716685/175832153-d1966287-44ab-4ed2-a9ef-fed2cf6091ad.mp4

In conclusion it helps, but not by much.
I still have to optimize the creation of a 'Date()' object on each frame, but it's not really expensive haha

**Comment by ManApart** [comment]
Haha, thanks. Well I appreciate it either way. At some point I'll get the motivation up to pause other projects and take another pass at it as well.
